2.train-VAE/betatcvae.py

```diff
@@ -0,0 +1,300 @@
+            anneal_rate * self.gamma * kld_loss)
+        
+        return loss
+
+
+
+
+    
+def train_model(model, train_loader, optimizer):
+    model.train()
+    train_loss = 0
+    for batch in train_loader:
+        data = batch[0]
+        optimizer.zero_grad()
+        recon_batch, mu, log_var, z = model(data)
+        loss = model.loss_function(recon_batch, data, mu, log_var, z)
+        loss.backward()
+        train_loss += loss.item()
+        optimizer.step()
+    avg_train_loss = train_loss / len(train_loader.dataset)
+    return avg_train_loss
+
+def train_vae(model, train_loader, optimizer, epochs):
+    train_loss_history = []
+    model.train()
+    for epoch in range(epochs):
+        avg_train_loss = train_model(model, train_loader, optimizer)
+        train_loss_history.append(avg_train_loss)
+        print(f"Epoch {epoch}, Loss: {avg_train_loss}")
+    return train_loss_history
+
+def evaluate_vae(model, val_loader):
+    model.eval()
+    val_loss = 0
+    with torch.no_grad():
+        for batch in val_loader:
+            data = batch[0]
+            recon, mu, log_var, z = model(data)
+            z = model.reparameterize(mu, log_var)
+            val_loss += model.loss_function(recon, data, mu, log_var, z).item()
+    return val_loss / len(val_loader.dataset)
+
+def compile_vae(model, train_loader, val_loader, test_loader, optimizer, epochs):
+    model.train()
+    train_loss_history = []
+    val_loss_history = []
+    test_loss_history = []
+
+    for epoch in range(epochs):
+        avg_train_loss = train_model(model, train_loader, optimizer)
+        train_loss_history.append(avg_train_loss)
+
+        avg_val_loss = evaluate_vae(model, val_loader)
+        val_loss_history.append(avg_val_loss)
+
+        avg_test_loss = evaluate_vae(model, test_loader)
+        test_loss_history.append(avg_test_loss)
+
+        print(f"Epoch {epoch+1}, Train Loss: {avg_train_loss}, Val Loss: {avg_val_loss}, Test Loss: {avg_test_loss}")
+
+    return train_loss_history, val_loss_history, test_loss_history
+
+def extract_latent_dimensions(model, data_loader, metadata):
+    model.eval()
+    latent_space = []
+    with torch.no_grad():
+        for batch in data_loader:
+            data = batch[1]
+            mu, log_var = model.encode(data)  # Extract the mean and log variance
+            latent_space.append(mu.cpu().numpy())  # Append the mean to the latent space
+    latent_space = np.concatenate(latent_space, axis=0)
+    latent_df = pd.DataFrame(latent_space)
+    latent_df.insert(0, 'ModelID', metadata['ModelID'])
+    latent_df_dir = pathlib.Path("./results/latent_df_tc.parquet")
+    latent_df.to_parquet(latent_df_dir, index=False)
+    return latent_df
+
+def weights(model, subset_train_df):
+    # Access the first layer of the encoder within the Sequential container
+    weight_matrix = model.encoder[0][0].weight.detach().cpu().numpy().T  # Extract weights from the first linear layer
```

Comment:
Is there a way to make this programmatic to extract weights from other layers too?